src/analysis/mean_inequalities.lean

```diff
@@ -103,47 +108,76 @@ begin
-    { rw [nnreal.div_def, nnreal.div_def], ring } },
-  { exact ne_of_gt (lt_trans zero_lt_one hq) },
-  { exact ne_of_gt (lt_trans zero_lt_one hp) }
-end
+end nnreal
 
-/-- Young's inequality, `ℝ` version -/
-theorem real.young_inequality {a b : ℝ} (ha : 0 ≤ a) (hb : 0 ≤ b)
-  {p q : ℝ} (hp : 1 < p) (hq : 1 < q) (hpq : 1/p + 1/q = 1) :
-  a * b ≤ a^p / p + b^q / q :=
-@nnreal.young_inequality ⟨a, ha⟩ ⟨b, hb⟩ ⟨p, le_trans zero_le_one (le_of_lt hp)⟩
-  ⟨q, le_trans zero_le_one (le_of_lt hq)⟩ hp hq (nnreal.coe_eq.1 hpq)
+namespace real
+
+theorem am_gm2_weighted {w₁ w₂ p₁ p₂ : ℝ} (hw₁ : 0 ≤ w₁) (hw₂ : 0 ≤ w₂)
+  (hp₁ : 0 ≤ p₁) (hp₂ : 0 ≤ p₂) (hw : w₁ + w₂ = 1) :
+  p₁ ^ w₁ * p₂ ^ w₂ ≤ w₁ * p₁ + w₂ * p₂ :=
+nnreal.am_gm2_weighted ⟨w₁, hw₁⟩ ⟨w₂, hw₂⟩ ⟨p₁, hp₁⟩ ⟨p₂, hp₂⟩ $ nnreal.coe_eq.1 $ by assumption
 
-theorem real.pow_am_le_am_pow (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
+/-- Young's inequality, a version for nonnegative real numbers. -/
+theorem young_inequality_of_nonneg {a b p q : ℝ} (ha : 0 ≤ a) (hb : 0 ≤ b)
+  (hpq : p.is_conjugate_exponent q) :
+  a * b ≤ a^p / p + b^q / q :=
+by simpa [← rpow_mul, ha, hb, hpq.ne_zero, hpq.symm.ne_zero, div_eq_inv_mul]
+  using real.am_gm2_weighted (le_of_lt hpq.one_div_pos) (le_of_lt hpq.symm.one_div_pos)
+    (rpow_nonneg_of_nonneg ha p) (rpow_nonneg_of_nonneg hb q) hpq.inv_add_inv_conj
+
+/-- Young's inequality, a version for arbitrary real numbers. -/
+theorem young_inequality (a b : ℝ) {p q : ℝ} (hpq : p.is_conjugate_exponent q) :
+  a * b ≤ (abs a)^p / p + (abs b)^q / q :=
+calc a * b ≤ abs (a * b)                   : le_abs_self (a * b)
+       ... = abs a * abs b                 : abs_mul a b
+       ... ≤ (abs a)^p / p + (abs b)^q / q :
+  real.young_inequality_of_nonneg (abs_nonneg a) (abs_nonneg b) hpq
+
+theorem pow_am_le_am_pow (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
   (hw' : ∑ i in s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) (n : ℕ) :
   (∑ i in s, w i * z i) ^ n ≤ ∑ i in s, (w i * z i ^ n) :=
 (convex_on_pow n).map_sum_le hw hw' hz
 
-theorem nnreal.pow_am_le_am_pow (w z : ι → ℝ≥0) (hw' : ∑ i in s, w i = 1) (n : ℕ) :
+theorem pow_am_le_am_pow_of_even (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
+  (hw' : ∑ i in s, w i = 1) {n : ℕ} (hn : n.even) :
+  (∑ i in s, w i * z i) ^ n ≤ ∑ i in s, (w i * z i ^ n) :=
+(convex_on_pow_of_even hn).map_sum_le hw hw' (λ _ _, trivial)
+
+theorem fpow_am_le_am_fpow (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
+  (hw' : ∑ i in s, w i = 1) (hz : ∀ i ∈ s, 0 < z i) (m : ℤ) :
+  (∑ i in s, w i * z i) ^ m ≤ ∑ i in s, (w i * z i ^ m) :=
+(convex_on_fpow m).map_sum_le hw hw' hz
+
+theorem rpow_am_le_am_rpow (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
+  (hw' : ∑ i in s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) {p : ℝ} (hp : 1 ≤ p) :
+  (∑ i in s, w i * z i) ^ p ≤ ∑ i in s, (w i * z i ^ p) :=
+(convex_on_rpow hp).map_sum_le hw hw' hz
+
+end real
+
+namespace nnreal
+
+/-- Young's inequality, `ℝ≥0` version -/
+theorem young_inequality (a b : ℝ≥0) {p q : ℝ≥0} (hp : 1 < p) (hpq : 1 / p + 1 / q = 1) :
```

Comment:
What's the point of having nnreal exponents, instead of conjugate real exponents here?

Reply:
In the RHS I divide by `p` and `q`. With real exponent I would need something like `nnreal.of_real` or `⟨p, hp⟩`.

Reply:
You're right.